### PR TITLE
Merge `package:source_gen`

### PR DESCRIPTION
**Pull request by mosuem**
- [x] Move and fix workflow files, labeler.yaml, and badges in the README.md
- [x] Rev the version of the package, so that pub.dev points to the correct site
- [x] Add a line to the changelog:
```
* Move to `dart-lang/build` monorepo.
```

- [x] Add the package to the top-level readme of the monorepo:
```
| [source_gen](source_gen/source_gen/) |  | [![package issues](https://img.shields.io/badge/package:source_gen-4774bc)](https://github.com/dart-lang/tools/issues?q=is%3Aissue+is%3Aopen+label%3Apackage%3Asource_gen) | [![pub package](https://img.shields.io/pub/v/source_gen.svg)](https://pub.dev/packages/source_gen) |
```

- [ ] **Important!** Merge the PR with 'Create a merge commit' (enabling then disabling the `Allow merge commits` admin setting)
- [ ] Update the auto-publishing settings on https://pub.dev/packages/source_gen/admin
- [ ] Add the following text to https://github.com/dart-lang/source_gen/:'

```
> [!IMPORTANT]  
> This repo has moved to https://github.com/dart-lang/build/tree/master/source_gen/source_gen
```

- [ ] Publish using the autopublish workflow
- [ ] Push tags to GitHub using
```
git tag --list 'source_gen*' | xargs git push origin
```

- [ ] Close open PRs in dart-lang/source_gen with the following message:

```
Closing as the [dart-lang/source_gen](https://github.com/dart-lang/source_gen) repository is merged into the [dart-lang/build](https://github.com/dart-lang/build) monorepo. Please re-open this PR there!
```
      
- [ ] Transfer issues by running
```
dart run pkgs/repo_manage/bin/report.dart transfer-issues --source-repo dart-lang/source_gen --target-repo dart-lang/build --add-label package:source_gen --apply-changes
```

- [ ] Archive https://github.com/dart-lang/source_gen/

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: source_gen/tool/ci.sh [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#!/bin/bash
```
Comment (jakemac53) [inline source_gen/tool/ci.sh:1]: This script can be deleted, it was a part of the old CI setup

File: source_gen/.gitignore [hunk 2]
```diff
@@ -0,0 +1,5 @@
+.dart_tool
+.packages
+.pub
+packages
+pubspec.lock
```
Comment (kevmoo) [inline source_gen/.gitignore:5]:
```suggestion
.dart_tool
pubspec.lock
```

This can be cleaner 😄

**Comment by kevmoo** [comment]
I know it's a bit ugly, but I really like the mono_repo-ness of source_gen. It makes it MUCH easier to explain how to use it

I think I'd argue this doesn't make sense to merge as a mono-repo.

**Comment by kevmoo** [comment]
@natebosch – nested packages give me a headache.

**Comment by jakemac53** [comment]
From the other perspective, what do we actually gain by merging this into the mono repo? This has never been clear to me. 

Here are the downsides from my perspective:

- Cannot watch only issues pertaining to source_gen (or the opposite, people who care about build but not source_gen can't only watch the build related issues).
- More complexity around GitHub actions, and it's wasteful in terms of CI resources as currently set up.
- More steps to set up a "clean" workspace in VsCode (opening up just the root of the build or source_gen repos was quite reasonable before, now there is going to be more there, and most of it isn't relevant to a given task).
- Larger issue counts, larger PR counts, which are difficult to manage. I think having separate repos is much nicer in this regard.
- Can use pub workspaces more easily, these don't really work well with unrelated packages.

**Comment by mosuem** [comment]
> From the other perspective, what do we actually gain by merging this into the mono repo? This has never been clear to me.
> 
> Here are the downsides from my perspective:
> 
> * Cannot watch only issues pertaining to source_gen (or the opposite, people who care about build but not source_gen can't only watch the build related issues).

This is indeed a downside for users of single packages. From a team and maintenance perspective on the other hand, we saw many issues and PRs fall through the cracks as people didn't follow some repositories. Seeing all issues and PRs in one place makes this simpler. We have tried to ease this in other repos by providing links to filtered lists, see the https://github.com/dart-lang/tools/ readme.

> * More complexity around GitHub actions, and it's wasteful in terms of CI resources as currently set up.

I don't think the CI is wasteful - through path restrictions, CI should only run for the changed packages. Also, we use very little CI capacity across the dart-lang org. The CI setup also becomes simpler IMO, as we don't have to update and maintain CI scripts across different repositories, for example the PR health workflow. 

> * More steps to set up a "clean" workspace in VsCode (opening up just the root of the build or source_gen repos was quite reasonable before, now there is going to be more there, and most of it isn't relevant to a given task).

I agree with this, I don't know if this is a big issue in practice. We could add some documentation on https://git-scm.com/docs/git-sparse-checkout to help a bit.

> * Larger issue counts, larger PR counts, which are difficult to manage. I think having separate repos is much nicer in this regard. Labels are not really equivalent.

I agree with this problem in principle, but don't think this makes it much harder to manage.

> * Can't really use pub workspaces, they don't really work well with unrelated packages.

You can use multiple pub workspaces if needed, as is happening here.

We should probably discuss this elsewhere, such as https://github.com/dart-lang/sdk/issues/56591, or do you mean those points only in regard to this merge in particular? FWIW, I think monorepos make everything easier and am a big fan, if it were up to me we would have only one repository for all packages ;)

**Comment by kevmoo** [comment]
I'd rather have a seperate pub workspace then. Although, honestly, I'd strongly vote to just leave this stand-alone.

**Review by kevmoo — CHANGES_REQUESTED** [review]
We agreed to drop this one, right?

**Comment by mosuem** [comment]
> We agreed to drop this one, right?

Not that I know of - my understanding is that the last state of the offline discussion is that @devoncarew syncs with @jakemac53 about the state of this.

**Comment by devoncarew** [comment]
To update, I did sync w/ @jakemac53 about this.

First, as an aside, we (the ecosystem team) are generally sold on the economics of maintaining multiple packages in fewer repos (it's cheaper), and, that there's a quality benefit (ability to invest more in tooling, fewer repos to watch, fewer PR queues to know about). Owning 100 repos is strictly more expensive than owning a dozen. However, I do recognize that there are trade-offs, especially with repos that get a high number of issues reported (in our space, that's generally the flutter/flutter and dart-lang/sdk repos).

Having said that, I want to mostly focus this PR on merging source_gen or not, not a general discussion of the value of monorepoization.

My stance if that if a package is currently well-owned, and the owner would prefer their current setup, we're not going to make changes. `dart-lang/build` is currently well-owned, and Jake would prefer to keep `source_gen` separate from it. We don't need to move source_gen into build. Separately, Kevin owns and maintains source_gen and plans to continue that; he prefers it in its own repo, so,  we won't look to move it into a something like `dart-lang/tools`.

I think source_gen stays as-is as we can close this PR. @mosuem - thanks for creating the PR (and for the work in consolidating packages generally :) ).

I'll make a brief why-we're doing-this comment on https://github.com/dart-lang/sdk/issues/56591. I believe our current round of consolidation (Moritz's open PRs) will finish off our planned arc of work here.

**Pull request closed without merging**